### ISSUE TITLE
SignalR .NET client deadlocks when calling StopAsync inside On handler

### ISSUE THREAD

**Issue by andre-ss6** [dotnet/aspnetcore#19750]
### Describe the bug
If `HubConnection.StopAsync` is called inside an `On` handler from the .NET client it will never return. 

During my testing, I've also noticed that if any other thread calls `StopAsync` from somewhere else during this period, it will also never return and the cancellation token will not fire, even if cancellation has been requested.

Edit: included the scenario described above.

#### Server
```cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.SignalR;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

namespace SignalRDeadlockReproduce
{
    public class FooHub : Hub
    {
        public FooHub(IHubContext<FooHub> hubContext)
        {
            var t = new Task(async () =>
            {
                await Task.Delay(TimeSpan.FromSeconds(5));
                await hubContext.Clients.All.SendAsync("deadlock");
            }, TaskCreationOptions.LongRunning);
            t.Start();
        }
    }

    public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddSignalR();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseRouting();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapHub<FooHub>("/foo");
            });
        }
    }
}
```

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.SignalR" Version="1.1.0" />
  </ItemGroup>

</Project>
```

#### Client
```cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR.Client;
using static System.Console;

namespace SignalRDeadlockReproduce.Console
{
    class Program
    {
        static async Task Main(string[] args)
        {
            var conn = new HubConnectionBuilder().WithUrl("http://localhost:54841/foo").Build();
            conn.On("deadlock", async () =>
            {
                WriteLine("Here it comes");
                await conn.StopAsync();
                WriteLine("Done! Press any key to exit..."); // Will never print
            });

            await conn.StartAsync();
            WriteLine("Connection open...");

            await Task.Delay(TimeSpan.FromSeconds(10));
            new Thread(async () =>
            {
                WriteLine("Me too!");
                var cts = new CancellationTokenSource();
                cts.CancelAfter(500);
                try
                {
                    await conn.StopAsync(cts.Token);
                }
                catch (Exception)
                {
                    WriteLine("Cancelled"); // Will never print
                }
                WriteLine("Yay! I also deadlocks :)"); // Will never print
            }).Start();

            ReadKey();
        }
    }
}

```

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.SignalR.Client" Version="3.1.2" />
  </ItemGroup>

</Project>
```

I use the background task to emulate the currently suggested workaround to closing connections server-side outside of hubs (which is calling a `KindlyCloseYourConnection` method on the client, since `IHubContext` doesn't offer a way to `Abort` connections). The same thing happens if the client `InvokeAsync` a method that eventually calls `deadlock` though.

### Further technical details
- ASP.NET Core 3.1, SignalR latest version for both client and server
- `dotnet --info`:
```
SDK do .NET Core (refletindo qualquer global.json):
 Version:   3.1.101
 Commit:    b377529961

Ambiente de runtime:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.101\

Host (useful for support):
  Version: 3.1.2
  Commit:  916b5cba26

.NET Core SDKs installed:
  3.1.101 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

- VS 16.4.5

**Comment by davidfowl** [comment]
Might be related to https://github.com/dotnet/aspnetcore/issues/17584, not sure if StopAsync is different, we'll need to take a look.

**Comment by anurse** [comment]
Probably true, but also probably a different code path so we'll need to make a different fix.

**Comment by BrennanConroy** [comment]
It's the same code. Stop will wait for the receive loop to complete, but the receive loop waits on the .On handlers
https://github.com/dotnet/aspnetcore/blob/6912dda560f52f761d9528dba786a7872ad1739f/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L1223
which is where Stop is being called from.

I can see two changes:
1. We don't wait for user methods and just `_ = StartProcessingInvocationMessages(...)`
https://github.com/dotnet/aspnetcore/blob/6912dda560f52f761d9528dba786a7872ad1739f/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L1126
1. Put a time limit on how long we wait either at the end of the receive loop or in Stop

**Comment by dgroh** [comment]
This is still on issue on 3.1.3 right? 

Actually I wanted to await for the connection to stop so I do some logic...

What I'm doing right now is:

`_connection.Closed += OnConnection_Closed;`

```csharp
        private Task OnConnection_Closed(Exception arg)
        {
            // TODO: My logic here
        }
```

Does this make sense?
